In [1]:
import pandas as pd
dataset = pd.read_csv("Data_for_UCI_named.csv")
dataset.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

In [2]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    10000 non-null  float64
 1   tau2    10000 non-null  float64
 2   tau3    10000 non-null  float64
 3   tau4    10000 non-null  float64
 4   p1      10000 non-null  float64
 5   p2      10000 non-null  float64
 6   p3      10000 non-null  float64
 7   p4      10000 non-null  float64
 8   g1      10000 non-null  float64
 9   g2      10000 non-null  float64
 10  g3      10000 non-null  float64
 11  g4      10000 non-null  float64
 12  stab    10000 non-null  float64
 13  stabf   10000 non-null  object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [3]:
target_frame = dataset["stabf"]
target_frame.head()

0    unstable
1      stable
2    unstable
3    unstable
4    unstable
Name: stabf, dtype: object

In [4]:
dataset.drop(columns='stabf',axis=1, inplace=True)
dataset.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860

In [5]:
from sklearn.preprocessing import StandardScaler
standardScaler = StandardScaler()
standardScaler.fit(dataset)
scaled_dataset = pd.DataFrame(standardScaler.transform(dataset), columns=dataset.columns.values)
scaled_dataset.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0 -0.835374 -0.791317  1.141704  1.652103  0.017397  1.079405 -0.017078   
1  1.478297 -0.126705 -0.803111 -1.415043  1.752124 -1.593619 -1.438158   
2  1.357093  1.312140 -0.803499 -1.471504 -0.458492  0.098253 -0.062840   
3 -1.653138  0.882289 -0.278354 -1.060901  0.284250  0.513904 -1.591046   
4 -0.771543  0.860108 -0.111670  1.680114 -0.298075  0.287450 -1.376343   

         p4        g1        g2        g3        g4      stab  
0 -1.092545  0.457467  1.220013  1.321628  1.579026  1.073120  
1 -0.011575 -0.406791  1.230354  0.135424  0.936256 -0.587487  
2  0.760963 -1.319852  0.881299  1.146596 -1.513802 -0.332095  
3  0.583414 -0.287304  1.647250  1.474543 -0.591750  0.355922  
4  1.606636  0.992226 -0.253610  0.481133  1.079063  0.924487

# 1) Download Electrical Grid Stability Simulated Datadatasets. Classify "Electrical Grid Stability Simulated Data" classes using KNN. Use the same data splitting and performance metrics that you have used in previous week (week 7, Q-2). Report your findings including comparison of results with week 7

In [6]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, f1_score
## Same method used in Week 7 Exercise.
def print_model_metrics(true_labels, predicted_labels):
    # Calculate metrics for classification problems
    acc = accuracy_score(true_labels, predicted_labels)
    conf_matrix = confusion_matrix(true_labels, predicted_labels)
    prec = precision_score(true_labels, predicted_labels, average='macro')
    f1 = f1_score(true_labels, predicted_labels, average='macro') # for multiclass problems.,→ Use 'binary' for binary problems
    # Print metrics
    print("Accuracy:", acc)
    print("\nConfusion Matrix:\n", conf_matrix)
    print("\nPrecision:", prec)
    print("\nF1 Score:", f1)

I am Also assuming we should also do the PCA as in exercise 7 to be able to properly compare the results.

In [7]:
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
pca=PCA(n_components=3)


principal_components = pca.fit_transform(scaled_dataset)

In [9]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split ( principal_components, target_frame, test_size=0.2 )
classifier = KNeighborsClassifier().fit(X_train,y_train.values.ravel())
predicted = classifier.predict(X_test)
print_model_metrics(y_test, predicted)

Accuracy: 0.9025

Confusion Matrix:
 [[ 595  110]
 [  85 1210]]

Precision: 0.8958333333333333

F1 Score: 0.8923179932491665


In our results, calculating the result of KNN Classification with DEFAULT HYPERPARAMETERS, we had an accuracy of .9025 vs .909 of Week's 7 resut.
Our F1 Score = .892 vs .897 and Our precision was .89 vs .92 from last week. Our results also show a little bit of mixed results comparing the confusion matrix, 1210 vs 1250 true positives, 595 vs 568 true negatives, 110 vs 151 false positives and 85 vs 31 false negatives. Overall, last week's algorithm was slightly better to the given problem, even though the difference did not show to be too big. The current model still show a slightly better metric to avoid false positives though.


# 2)Load Electrical Grid Stability Simulated Data and create classification model using DT algorithm using 50-50% and 80-20% data splitting methods. Compare performances of these two models and explain impact of difference in data splitting on the performances of the model.


In [10]:
##Data Already loaded under dataset. Previous distribution was 80-20: Running tree:
from sklearn.tree import DecisionTreeClassifier
treeclassifier = DecisionTreeClassifier( random_state=13)
treeclassifier.fit(X_train,y_train.values.ravel())
tree_predicted = treeclassifier.predict(X_test)
print_model_metrics(y_test, tree_predicted)


Accuracy: 0.8805

Confusion Matrix:
 [[ 591  114]
 [ 125 1170]]

Precision: 0.868316973842218

F1 Score: 0.8695685036860541


In [11]:
X_train_50, X_test_50, y_train_50, y_test_50 = train_test_split ( principal_components, target_frame, test_size=0.5)
treeclassifier_50 = DecisionTreeClassifier( random_state=13)
treeclassifier_50.fit(X_train_50,y_train_50.values.ravel())
tree_predicted_50 = treeclassifier.predict(X_test_50)
print_model_metrics(y_test_50, tree_predicted_50)

Accuracy: 0.9744

Confusion Matrix:
 [[1777   64]
 [  64 3095]]

Precision: 0.9724883544063941

F1 Score: 0.9724883544063941


In our result we can infer that all the metrics were superior in the 50/50 split. That is a strong indication that if we do a 80/20 split of tha data we are incurring into overfitting, degrading the overall result of our predictions.

# 3) Create two more KNN-based classification models using the dataset used in Q1 by varying distance metrics such as using  cityblock and manhattan. Report the performances of the developed models including Q1 and explain the similarity or differences if any.

In [12]:
classifier_manhattan = KNeighborsClassifier(metric="manhattan").fit(X_train,y_train.values.ravel())

classifier_cityblock = KNeighborsClassifier(metric="cityblock").fit(X_train,y_train.values.ravel())

predicted_manhattan = classifier_manhattan.predict(X_test)
predicted_city = classifier_cityblock.predict(X_test)

print_model_metrics(y_test, predicted_manhattan)
print_model_metrics(y_test, predicted_city)



Accuracy: 0.9015

Confusion Matrix:
 [[ 596  109]
 [  88 1207]]

Precision: 0.8942591407597007

F1 Score: 0.8913606635712727
Accuracy: 0.9015

Confusion Matrix:
 [[ 596  109]
 [  88 1207]]

Precision: 0.8942591407597007

F1 Score: 0.8913606635712727


As a result, the scores were quite similar in the Q1 we got:

Accuracy: 0.9025
Confusion Matrix:
 [[ 595  110]
 [  85 1210]]
Precision: 0.8958333333333333
F1 Score: 0.8923179932491665

for BOTH the manhattan and the cityblock metrics we got:
Accuracy: 0.9015

Confusion Matrix:
 [[ 596  109]
 [  88 1207]]

Precision: 0.8942591407597007

F1 Score: 0.8913606635712727
The result of manhattan and cityblock were identical. upon further research, i found reference (https://stackoverflow.com/questions/37303146/distance-metrics-in-scikit-learn) that both the metrics actually point to the same distance calculation implementation, being in fact, the same. As a conclusion, the distance calculation methods used in the model, in this case did not have a major impact on the predictions.